GH-1881 - Fix interface method attributes when generating proxies

PleasantD opened the pull request:
See #1881

fredericDelaporte commented:
Some tests are failing. But also, there are no new tests added. A test should be added, demonstrating the current failure this PR aims at fixing.

PleasantD commented:
Added the necessary test.
Also fixed the other tests that were failing, which included fixing an assumption about creating a proxy from an interface and fixing the IL generated for explicit implementations of interface methods.

fredericDelaporte commented:
The `TuplizerDynamicEntity.It` test fails with Firebird, Oracle and SQL Server CE (Compact Edition). The five failing builds are failed due to this test.

fredericDelaporte commented on src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs:
```diff
@@ -118,13 +118,13 @@ private TypeInfo CreateUncachedProxyType(System.Type baseType, IReadOnlyCollecti
 
 			// Make the proxy serializable
-			AddSerializationSupport(baseType, baseInterfaces, typeBuilder, interceptorField, defaultConstructor);
+			AddSerializationSupport(parentType, baseInterfaces, typeBuilder, interceptorField, defaultConstructor);
```
This one must stay `baseType`, otherwise when deserializing a proxy with a base type being an interface, it will create a new proxy type with `baseType` `DummyProxy` instead of the interface, causing proxy type caching issues. (Two types for the same proxy may be built.)

fredericDelaporte commented on src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs:
```diff
@@ -118,13 +118,13 @@ private TypeInfo CreateUncachedProxyType(System.Type baseType, IReadOnlyCollecti
 			// Provide a custom implementation of ISerializable
 			// instead of redirecting it back to the interceptor
-			foreach (MethodInfo method in ProxyBuilderHelper.GetProxiableMethods(baseType, interfaces).Where(method => method.DeclaringType != typeof(ISerializable)))
+			foreach (MethodInfo method in ProxyBuilderHelper.GetProxiableMethods(parentType, interfaces).Where(method => method.DeclaringType != typeof(ISerializable)))
```
Ok, I have rechecked the TeamCity builds, seen what was the error, and seen I have missed checking that test in `DynamicProxyTests`. So now I see what it does fix. This trouble does not apply to the static proxy case indeed.

Now this change also seems to be what causes `TuplizerDynamicEntity.It` failure with some databases. (I have not dug further yet.)

PleasantD commented on src/NHibernate.Test/NHSpecificTest/GH1881/TestFixture.cs:
```diff
@@ -0,0 +1,33 @@
+		[Test]
+		public void InterfacesShouldBeImplementedExplicitlyOnProxies()
+		{
+			var proxy = new ProxyFactory().CreateProxy(typeof(PersistentGenericBag<object>), null, typeof(ILazyInitializedCollection));
```
Well... no, but that's kind of the point.
To reproduce the issue I would need a type that isn't validatable.

```
public interface IFoo
{
    bool Prop { get; }
}

public class Bar : IFoo
{
    public bool Prop => false;
}

// throws the same error
new ProxyFactory().CreateProxy(typeof(Bar), null, typeof(IFoo));
```

The proxy factory generates a proxy type for `Bar`, skipping `Prop` because it isn't proxyable.
But it adds the property from `IFoo`, which is proxyable, which is wrong and causes errors.
With the explicit implementation of the interface the proxy is valid for the interface, if not for the base type.

fredericDelaporte replied:
It would be still better to use this type for the test than this big thing that `PersistentGenericBag` is.

So, there are actually no way to reproduce the failure when using the dynamic proxy factory only through NHibernate's proxy handling, isn't it? (Or does Envers disable the default proxy validation in NHibernate? There is a setting allowing this.)

So this change looks needed for somewhat external users of the dynamic proxy factory, using it for their own needs with less strict prerequisites on proxied classes. But that is not the purpose of NHibernate to provide a general purpose proxifying mechanism. In this spirit the new static proxy factory API is way less exposed (more internal), and the old dynamic proxy should be gone in 6.0.

Maybe Envers will need to use an external proxy library rather than using the NHibernate internal one. (Unless it can be changed to use the static proxy factory, if it is not too specific to NHibernate for being used for Envers case. But I doubt about this.)

(In other words, this all looks to me at lot like this other [discussion](https://github.com/nhibernate/nhibernate-core/pull/1709#commitcomment-29603179).)

hazzik replied:
In the example above `Bar` will fail the validation.

hazzik replied:
> Or does Envers disable the default proxy validation in NHibernate? There is a setting allowing this.

They use it for something different - as a general purpose proxy factory to inject an interceptor, not for nhibernate entities.

fredericDelaporte commented:
By the way, about the commit message, since the switch to GitHub, we have change the expected formalism.

Just follow usual guidelines now, like using on first line a short dot-less sentence starting with an imperative verb.

No need to reference the issue, when the PR is itself the issue. Otherwise, reference it on subsequent lines, with GitHub convention rather than the `GH` prefix (which is there just for having a valid folder name in NHSpecificTest, when tests have to be added there rather than into a feature folder).

So here for example, the commit message could be:

```
Fix interface method attributes when generating proxies

Fix #1881
```

PleasantD commented on src/NHibernate/Proxy/NHibernateProxyBuilder.cs:
```diff
@@ -394,7 +394,7 @@ private static void ImplementGetLazyInitializer(TypeBuilder typeBuilder, MethodI
 
 			IL.Emit(OpCodes.Ldarg_0);
-			EmitCallMethod(IL, OpCodes.Call, method);
+			EmitCallMethod(IL, method.DeclaringType.IsInterface ? OpCodes.Callvirt : OpCodes.Call, method);
```
I added the test and it fails

```
NHibernate.HibernateException : Creating a proxy instance failed
  ----> System.ArgumentException : Expression of type 'PersistentGenericBag`1Proxy' cannot be used for return type 'NHibernate.Proxy.INHibernateProxy'
   at NHibernate.Proxy.StaticProxyFactory.GetProxy(Object id, ISessionImplementor session) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 31
   at NHibernate.Test.StaticProxyTest.StaticProxyFactoryFixture.VerifyInterfacesImplementedExplicitlyOnProxies() in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate.Test\StaticProxyTest\StaticProxyFactoryFixture.cs:line 153
--ArgumentException
   at System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, Expression& body, ReadOnlyCollection`1 parameters, String paramName)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, String name, Boolean tailCall, IEnumerable`1 parameters)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Boolean tailCall, IEnumerable`1 parameters)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, ParameterExpression[] parameters)
   at NHibernate.Proxy.StaticProxyFactory.CreateProxyActivator(ProxyCacheEntry pke) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 42
   at NHibernate.Proxy.StaticProxyFactory.<GetProxy>b__3_0(ProxyCacheEntry pke) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 23
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at NHibernate.Proxy.StaticProxyFactory.GetProxy(Object id, ISessionImplementor session) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 23
```

PleasantD replied:
I will admit that I don't really know how to write the IL that would be needed for this.

The complete solution would be to have the explicit interface call `this.Method` while the class implementation calls `base.Method` if the base is non-abstract and returns `default` otherwise.

hazzik replied:
This is incorrect. `base` === `OpCodes.Call`.

hazzik replied:
@PleasantD I've fixed your concerns in #1884.

PleasantD commented on src/NHibernate.Test/NHSpecificTest/GH1881/TestFixture.cs:
```diff
@@ -0,0 +1,33 @@
+	public class TestFixture
+	{
+		[Test]
+		public void InterfacesShouldBeImplementedExplicitlyOnProxies()
```
I added the test and it fails

```
NHibernate.HibernateException : Creating a proxy instance failed
  ----> System.ArgumentException : Expression of type 'PersistentGenericBag`1Proxy' cannot be used for return type 'NHibernate.Proxy.INHibernateProxy'
   at NHibernate.Proxy.StaticProxyFactory.GetProxy(Object id, ISessionImplementor session) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 31
   at NHibernate.Test.StaticProxyTest.StaticProxyFactoryFixture.VerifyInterfacesImplementedExplicitlyOnProxies() in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate.Test\StaticProxyTest\StaticProxyFactoryFixture.cs:line 153
--ArgumentException
   at System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, Expression& body, ReadOnlyCollection`1 parameters, String paramName)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, String name, Boolean tailCall, IEnumerable`1 parameters)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Boolean tailCall, IEnumerable`1 parameters)
   at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, ParameterExpression[] parameters)
   at NHibernate.Proxy.StaticProxyFactory.CreateProxyActivator(ProxyCacheEntry pke) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 42
   at NHibernate.Proxy.StaticProxyFactory.<GetProxy>b__3_0(ProxyCacheEntry pke) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 23
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at NHibernate.Proxy.StaticProxyFactory.GetProxy(Object id, ISessionImplementor session) in D:\ExternalGit\nhibernate-core-pleasantd\src\NHibernate\Proxy\StaticProxyFactory.cs:line 23
```

fredericDelaporte replied:
Yes, that should be `new HashSet<System.Type> { typeof(ILazyInitializedCollection), typeof(INHibernateProxy) }` instead.

But well, as per the comment I have posted some seconds ago [here](https://github.com/nhibernate/nhibernate-core/pull/1882#discussion_r227509078), it is maybe useless to support this in the static proxy factory, which is way less a general purpose proxifying mechanism than the dynamic proxy, and which may be unsuitable for Envers need.

PleasantD commented:
I suppose as an alternative we could just fix `AbstractPersistentCollection`.

Envers is generating this error because it generates proxies of the collections so it can hook into `ILazyInitializedCollection`.
`AbstractPersistentCollection` implements `ILazyInitializedCollection` but doesn't implement `WasInitialized` as virtual.

If we add the `virtual` keyword then it should fix Envers' issue with creating proxies of collections.
https://github.com/nhibernate/nhibernate-core/blob/5c184527752b6e1fe7d40e91536dfdae300e287c/src/NHibernate/Collection/AbstractPersistentCollection.cs#L633

hazzik commented:
Agree
On Wed, 24 Oct 2018 at 10:47, Duncan Munro <notifications@github.com> wrote:

> I suppose as an alternative we could just fix AbstractPersistentCollection
> .
>
> Envers is generating this error because it generates proxies of the
> collections so it can hook into ILazyInitializedCollection.
> AbstractPersistentCollection implements ILazyInitializedCollection but
> doesn't implement WasInitialized as virtual.
>
> If we add the virtual keyword then it should fix Envers' issue with
> creating proxies of collections.
>
> https://github.com/nhibernate/nhibernate-core/blob/5c184527752b6e1fe7d40e91536dfdae300e287c/src/NHibernate/Collection/AbstractPersistentCollection.cs#L633
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nhibernate/nhibernate-core/pull/1882#issuecomment-432432908>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI1l3XgCURm2-fSLUf5BxWY6XdFtEG8ks5un45ogaJpZM4Xxhs9>
> .
>

PleasantD commented:
How would you like to proceed with this?

Please let me know what needs to be done to fix this up or if you'd rather just patch `AbstractPersistentCollection`

fredericDelaporte commented:
It seems patching `AbstractPersistentCollection` will be less controversial.

fredericDelaporte commented on src/NHibernate/Proxy/NHibernateProxyBuilder.cs:
```diff
@@ -398,6 +398,7 @@ private static void ImplementGetLazyInitializer(TypeBuilder typeBuilder, MethodI
 			if (method.DeclaringType.IsInterface &&
 			    method.DeclaringType.IsAssignableFrom(parentType))
 			{
+				Delegate
```
Hum, maybe your keyboard focus was not where you were expecting it to be when you typed this.

hazzik replied:
Yes

fredericDelaporte commented on src/NHibernate.Test/DynamicEntity/DataProxyHandler.cs:
```diff
@@ -53,6 +53,15 @@ public object Intercept(InvocationInfo info)
 			{
 				return GetHashCode();
 			}
+			else if ("Equals".Equals(methodName))
```
In other words, you suggest to simplify this test fix by just calling `RuntimeHelpers.Equals(this, obj)` instead of bothering calling a base implementation of a class we are not directly deriving from?

Ok.

hazzik commented on src/NHibernate/Proxy/ProxyBuilderHelper.cs:
```diff
@@ -137,7 +137,24 @@ internal static MethodBuilder GetObjectDataMethodBuilder(TypeBuilder typeBuilder
+					MethodAttributes.NewSlot |
+					MethodAttributes.SpecialName;
+				// .NET uses an expanded name for explicit interface implementation methods.
+				name = typeBuilder.FullName + "." + method.DeclaringType.FullName + "." + name;
```
This is not what .NET have for methods.

fredericDelaporte commented:
> I suppose as an alternative we could just patch `AbstractPersistentCollection`.
>
>Envers is generating this error because it generates proxies of the collections so it can hook into `ILazyInitializedCollection`.
>`AbstractPersistentCollection` implements `ILazyInitializedCollection` but doesn't implement `WasInitialized` as virtual.
>
>If we add the `virtual` keyword then it should fix Envers' issue with creating proxies of collections.
>https://github.com/nhibernate/nhibernate-core/blob/5c184527752b6e1fe7d40e91536dfdae300e287c/src/NHibernate/Collection/AbstractPersistentCollection.cs#L633

Looking into it, there are many more properties in `AbstractPersistentCollection` which are implicit non virtual interface implementations. They could trigger the same trouble, if their interface is included in the list of interfaces for the proxy to build.
The static proxy would include them if the proxy base type is the `AbstractPersistentCollection` class. The dynamic proxy would include them only if directly listed in the list of interfaces to implement.

So, proxified by the dynamic proxy, it may not trigger the issue. But would a proxy of `AbstractPersistentCollection` not also proxifying `IPersistentCollection` be actually valid? I do not think so. 

So implementing `WasInitialized` as virtual will just dodge one issue for then falling in less obvious ones: having calls to `IPersistentCollection` methods or properties hitting the base state instead of the delegated state. Or else Envers needs to include `IPersistentCollection` in the proxied interfaces list, and then many more changes are required in `AbstractPersistentCollection`.

fredericDelaporte commented:
For the same reason than #1881, closing.

>Closing because of following:
>
>1) Envers does not use the NHibernate's ProxyFactory anymore
>2) The proxy validator correctly identifies that the alleged type is not proxyable
